plugins/base/src/main/kotlin/transformers/pages/samples/SamplesTransformer.kt

```diff
@@ -55,19 +56,35 @@ abstract class SamplesTransformer(val context: DokkaContext) : PageTransformer {
         }
     }.toMap()
 
-    private fun ContentNode.addSample(contentPage: ContentPage, platform: PlatformData, fqName: String, analysis: Map<PlatformData, EnvironmentAndFacade>): ContentNode {
-        val facade = analysis[platform]?.facade ?:
-            return this.also { context.logger.warn("Cannot resolve facade for platform ${platform.name}")}
+    private fun ContentNode.addSample(contentPage: ContentPage, platform: PlatformData?, fqName: String, analysis: Map<PlatformData, EnvironmentAndFacade>): ContentNode {
+        val facade = if(platform == null) {
+            analysis.entries.find { it.key.platformType.name == "common" }?.value
+        } else {
+            analysis[platform]
+        }?.facade ?: return this.also { context.logger.warn("Cannot resolve facade for platform ${platform?.name ?: "expect"}") }
         val psiElement = fqNameToPsiElement(facade, fqName) ?:
             return this.also { context.logger.warn("Cannot find PsiElement corresponding to $fqName") }
         val imports = processImports(psiElement) // TODO: Process somehow imports. Maybe just attach them at the top of each body
         val body = processBody(psiElement)
-        val node = platformHintedContentCode(platform, contentPage.dri, body, "kotlin")
-        return this.safeAs<ContentGroup>()?.run { copy(
-            children = children.indexOfFirst { contentNode ->
-                contentNode.safeAs<ContentHeader>()?.children?.firstOrNull()?.safeAs<ContentText>()?.text == "Sample"
-            }.takeIf { it != -1 }?.let { children.apply { this.safeAs<MutableList<ContentNode>>()?.add(it+1, node) } } ?: children.also { context.logger.warn("Not found Sample block in ${contentPage.dri}")}
-        ) } ?: this.also { context.logger.warn("ContentPage ${contentPage.dri} cannot be cast to ContentGroup") }
+        val node = contentCode(contentPage.platforms(), contentPage.dri, body, "kotlin")
+
+        return bfs(fqName, node)
+    }
+
+    private fun ContentNode.bfs(fqName: String, node: ContentCode): ContentNode {
+        return when(this) {
+            is ContentHeader -> copy(children.map { it.bfs(fqName, node) })
+            is ContentCode -> copy(children.map { it.bfs(fqName, node) })
+            is ContentDRILink -> copy(children.map { it.bfs(fqName, node) })
+            is ContentResolvedLink -> copy(children.map { it.bfs(fqName, node) })
+            is ContentEmbeddedResource -> copy(children.map { it.bfs(fqName, node) })
+            is ContentTable -> copy(children = children.map { it.bfs(fqName, node) as ContentGroup })
+            is ContentList -> copy(children.map { it.bfs(fqName, node) })
+            is ContentGroup -> copy(children.map { it.bfs(fqName, node) })
+            is PlatformHintedContent -> copy(inner.bfs(fqName, node))
+            is ContentText -> if (text == fqName) node else this
+            else -> this
+        }
```

Comment:
Maybe most of these tags could point to `this`, but I'm not sure when one of these would be wrapper for some content in description

Reply:
Also it has a boilerplate, but I need smartcast for `copy` function